### PR TITLE
Fixed WithEndpoint to work without a protocol

### PR DESCRIPTION
**Pull request by Kaushal28**
Fixes: #3730 

When using the OTLP exporter, the WithEndpoint option requires an hostname, without the protocol.
However, the X_ENDPOINT environment variables requires an endpoint, including the protocol.

This was misleading, as both config have very similar names, but expect different values.

So, Fixed WithEndpoint to require a protocol, and be the same as the environment variable

### CODE REVIEW

**Comment by dmathieu** [comment]
I don't think we can just ship this kind of breaking change.
We should start by automatically adding a protocol if there are none instead of ignoring the value.

File: exporters/otlp/otlptrace/internal/otlpconfig/options.go [hunk 1]
```diff
@@ -244,11 +247,59 @@ func NewGRPCOption(fn func(cfg Config) Config) GRPCOption {
...
+	if cfg.Traces.Insecure {
+		return "http"
+	}
+	return "https"
+}
+
+// Maybe rawURL is of the form scheme:path.
+// (Scheme must be [a-zA-Z][a-zA-Z0-9+.-]*)
+// If so, return scheme; else return "".
+func hasScheme(rawURL string) bool {
```
Comment (Kaushal28) [inline exporters/otlp/otlptrace/internal/otlpconfig/options.go:279]: This is referenced from `net/url` package.

**Comment by Kaushal28** [comment]
Hey @dmathieu, I did changes as you mentioned. If that makes sense, I'll fix CI failures.

File: exporters/otlp/otlptrace/internal/otlpconfig/options_test.go [hunk 2]
```diff
@@ -92,10 +92,30 @@ func TestConfigs(t *testing.T) {
...
+		},
+		{
+			name: "Test With Endpoint with upper case scheme",
+			opts: []otlpconfig.GenericOption{
+				otlpconfig.WithEndpoint("HTTPS://127.0.0.1:4318"),
+			},
+			asserts: func(t *testing.T, c *otlpconfig.Config, grpcOption bool) {
+				assert.Equal(t, "127.0.0.1:4318", c.Traces.Endpoint)
+				assert.Equal(t, otlpconfig.DefaultTracesPath, c.Traces.URLPath)
 			},
```
Comment (dmathieu) [inline exporters/otlp/otlptrace/internal/otlpconfig/options_test.go:119]: Could you add a test with no scheme and the config set as insecure?

File: exporters/otlp/otlptrace/internal/otlpconfig/options.go [hunk 3]
```diff
@@ -244,11 +248,37 @@ func NewGRPCOption(fn func(cfg Config) Config) GRPCOption {
...
+func getScheme(cfg Config) string {
+	if cfg.Traces.Insecure {
+		return "http"
+	}
+	return "https"
+}
+
+// hasScheme determines if the given URL starts with a scheme.
+func hasScheme(rawURL string) bool {
+	return strings.HasPrefix(strings.ToLower(rawURL), "http://") || strings.HasPrefix(strings.ToLower(rawURL), "https://")
```
Comment (dmathieu) [inline exporters/otlp/otlptrace/internal/otlpconfig/options.go:279]: Nit, but we don't need to make the url lower twice.

```suggestion
	rawURL = strings.ToLower(rawURL)
	return strings.HasPrefix(rawURL, "http://") || strings.HasPrefix(rawURL, "https://")
```

File: exporters/otlp/otlptrace/internal/otlpconfig/options.go [hunk 4]
```diff
@@ -244,11 +248,37 @@ func NewGRPCOption(fn func(cfg Config) Config) GRPCOption {
 
 func WithEndpoint(endpoint string) GenericOption {
 	return newGenericOption(func(cfg Config) Config {
-		cfg.Traces.Endpoint = endpoint
+		// Add scheme if not present
+		if !hasScheme(endpoint) {
+			endpoint = (&url.URL{Scheme: getScheme(cfg), Host: endpoint}).String()
+		}
+		u, err := url.Parse(endpoint)
```
Comment (dmathieu) [inline exporters/otlp/otlptrace/internal/otlpconfig/options.go:255]: If no scheme is provided, we generate a new URL that we move to a string, and then we parse to a URL again.
Could we avoid doing that round twice by creating a `var u *url.URL` before the condition, and then setting it to a custom value, or parsing from the string?

File: exporters/otlp/otlptrace/internal/otlpconfig/options_test.go (hunk 2 above)
Comment (dmathieu) [reply exporters/otlp/otlptrace/internal/otlpconfig/options_test.go:119]: And one with an invalid URL? (adding a space at the beginning of the URL would make it unparsable for example)

**Comment by dmathieu** [comment]
When you're ready for it, I think you can mark this PR as ready for review :)

**Comment by Kaushal28** [comment]
Using `endpoint = fmt.Sprintf("%s://%s", getScheme(cfg), endpoint)` makes a unit test fail in `internal_logging_test.go`. I haven't changed anything specific to logging and also everything is passing in my local. Any ideas around this?

**Comment by dmathieu** [comment]
One other thing I just noticed: This change here is only for otlptrace. We should have consistently with otlpmetrics.

**Comment by Kaushal28** [comment]
Also did the similar changes for otlpmetrics.

**Review by dmathieu — APPROVED** [review]

File: exporters/otlp/otlpmetric/internal/oconf/options.go [hunk 5]
```diff
@@ -263,11 +265,32 @@ func NewGRPCOption(fn func(cfg Config) Config) GRPCOption {
...
+		// For OTLP/HTTP endpoint URLs without a per-signal
+		// configuration, the passed endpoint is used as a base URL
+		// and the signals are sent to these paths relative to that.
+		cfg.Metrics.URLPath = path.Join(u.Path, DefaultMetricsPath)
 		return cfg
 	})
 }
 
+// getScheme fetches scheme based on config.
+func getScheme(cfg Config) string {
```
Comment (Aneurysm9) [inline exporters/otlp/otlpmetric/internal/oconf/options.go:287]: I'm not sure this is a simple binary choice nor is it clear that `Insecure` necessarily translates to `http`.

**Review by Aneurysm9 — CHANGES_REQUESTED** [review]

File: exporters/otlp/otlpmetric/internal/oconf/options.go (hunk 5 above)
Comment (Kaushal28) [reply exporters/otlp/otlpmetric/internal/oconf/options.go:287]: Actually, I have referred that function from https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/otlptracehttp/client.go#L320

Comment (Kaushal28) [reply exporters/otlp/otlpmetric/internal/oconf/options.go:287]: Hey @Aneurysm9, Could you please take a look at my comment?

Comment (Aneurysm9) [reply exporters/otlp/otlpmetric/internal/oconf/options.go:287]: Not sure what happened to the link I attempted to put in that comment.  See https://github.com/grpc/grpc/blob/master/doc/naming.md.

File: exporters/otlp/otlpmetric/internal/oconf/options.go [hunk 6]
```diff
@@ -263,11 +265,32 @@ func NewGRPCOption(fn func(cfg Config) Config) GRPCOption {
...
+		u, err := url.Parse(endpoint)
+		if err != nil {
+			global.Error(err, "parse url", "input", endpoint)
+			return cfg
+		}
+		cfg.Metrics.Endpoint = u.Host
+		// For OTLP/HTTP endpoint URLs without a per-signal
+		// configuration, the passed endpoint is used as a base URL
+		// and the signals are sent to these paths relative to that.
+		cfg.Metrics.URLPath = path.Join(u.Path, DefaultMetricsPath)
```
Comment (Aneurysm9) [inline exporters/otlp/otlpmetric/internal/oconf/options.go:281]: I don't believe this is correct as gRPC endpoints are not HTTP URLs and should not be treated as such.  This is a general config option that applies to both gRPC and HTTP exporters and cannot include HTTP-specific behavior such as the addition of `DefaultMetricsPath`.

File: exporters/otlp/otlpmetric/internal/oconf/options.go (hunk 5 above)
Comment (pellared) [reply exporters/otlp/otlpmetric/internal/oconf/options.go:287]: @Aneurysm9 It looks OK. See https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/protocol/exporter.md


>  A scheme of `https` indicates a secure connection and takes precedence over the insecure configuration setting. A scheme of `http` indicates an insecure connection and takes precedence over the insecure configuration setting. If the gRPC client implementation does not support an endpoint with a scheme of http or https then the endpoint SHOULD be transformed to the most sensible format for that implementation.

**Review by pellared — CHANGES_REQUESTED** [review]
According the the spec, for gRPC protocol, neither `WithEndpoint` nor `OTEL_EXPORTER_OTLP_ENDPOINT` env var requires providing the scheme.

See https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/protocol/exporter.md

> **Insecure**: Whether to enable client transport security for the exporter's gRPC connection. This option only applies to OTLP/gRPC when an endpoint is provided without the `http` or `https` scheme 

The gRPC flavor should work without providing `http` or `https`.

**Comment by dmathieu** [comment]
While the gRPC spec does mention the exporter should work without an endpoint, it doesn't seem to mention it should not work with one.
This change support both (originally for backwards compatibility).

**Comment by Kaushal28** [comment]
Hey guys, I went through all the comments but currently I am not sure what comments are needed to be addressed as there seem to be multiple views.

**Comment by pellared** [comment]
> Hey guys, I went through all the comments but currently I am not sure what comments are needed to be addressed as there seem to be multiple views.

For sure this: https://github.com/open-telemetry/opentelemetry-go/pull/4016#discussion_r1176778775

**Comment by Kaushal28** [comment]
> For sure this: https://github.com/open-telemetry/opentelemetry-go/pull/4016#discussion_r1176778775

For this, I have a couple of doubts: 

1. What should be the default scheme if not it's not present? Should it be `dns` as mentioned [here](https://github.com/grpc/grpc/blob/master/doc/naming.md#:~:text=If%20no%20scheme%20prefix%20is%20specified%20or%20the%20scheme%20is%20unknown%2C%20the%20dns%20scheme%20is%20used%20by%20default.)? 
2. For metric endpoints, should I refer [this behaviour](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlpmetric/internal/oconf/envconfig.go#L73)? 

cc: @Aneurysm9

**Comment by pellared** [comment]
> 1. What should be the default scheme if not it's not present? Should it be `dns` as mentioned [here](https://github.com/grpc/grpc/blob/master/doc/naming.md#:~:text=If%20no%20scheme%20prefix%20is%20specified%20or%20the%20scheme%20is%20unknown%2C%20the%20dns%20scheme%20is%20used%20by%20default.)?

https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/protocol/exporter.md#configuration-options - therefore: `http`

**Comment by parauliya** [comment]
Hi @Aneurysm9/ @pellared ,
I would like to take this issue from here. Is that ok with you?

File: exporters/otlp/otlpmetric/internal/oconf/options.go (hunk 6 above)
Comment (parauliya) [reply exporters/otlp/otlpmetric/internal/oconf/options.go:281]: Hi @Aneurysm9 , @pellared ,
I am thinking to create two separate `WithEndPoint` functions for HTTP and gRPC and then we can put this code in the HTTP function. I understand about the changes from where this function is being called, will change it accordingly.
Does this seem ok with you?

Comment (parauliya) [reply exporters/otlp/otlpmetric/internal/oconf/options.go:281]: Hi @pellared , 
This is the change for `otlpmetric`. The same changes has been done for `otlptrace` as well. 
But there is no such comment provided in that section of the PR.
So just to confirm, Should I modify the same in `otlptrace` as well?

Comment (pellared) [reply exporters/otlp/otlpmetric/internal/oconf/options.go:281]: > Should I modify the same in `otlptrace` as well?

Yes 👍

**Comment by parauliya** [comment]
Hi @pellared ,
Could you please assign me this PR?
As I am not able to do the changes some trivial changes in the PR.

**Comment by pellared** [comment]
@parauliya Are you working on this one? Take notice that you need to sign the CLA.

**Comment by pellared** [comment]
Closing this PR as it is stale and also because I do not think we should make such change. More https://github.com/open-telemetry/opentelemetry-go/issues/3730#issuecomment-1770531895

**Pull request closed without merging**